### PR TITLE
Support waiting for the ACK response

### PR DESCRIPTION
**Pull request by BewareMyPower**
Fixes https://github.com/apache/pulsar-client-cpp/issues/102

### Modifications

Add a consumer configuration to set `ackReceiptEnabled`, if it's true, when sending a CommandAck request, generate the request id and set the `request_id` field so that the broker will respond with a `CommandAckResponse`. Here we pass the shared pointer to an atomic integer so that we no longer need to hold a weak pointer to `ClientImpl`.

Pass the user-defined callback to the ACK grouping tracker, when the `ackReceiptEnabled` is true, trigger the callback only after receiving the ACK response.

### Verifications

- Add `testAckReceiptEnabled` to verify when `ackReceiptEnabled` is true, the callback will be triggered after `ackGroupingMaxTime`.
- Support configuring `ackReceiptEnabled` for existing `TEST_P` parameterized tests in `AcknowledgeTest`.

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [ ] `doc-not-needed` 
(Please explain why)

- [x] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Comment by BewareMyPower** [comment]
Need to rebase after https://github.com/apache/pulsar-client-cpp/pull/231 is merged.

File: lib/Commands.cc
Comment (RobertIndie) [inline lib/Commands.cc:445]: The `ConsumerImpl. discardCorruptedMessage ` still passes the validationError to this function. And It will be incorrectly identified as the requestId here.
https://github.com/apache/pulsar-client-cpp/blob/14623ca6f5fbd84a4baf32a2c16673011d85f98c/lib/ConsumerImpl.cc#L828-L829

**Review by RobertIndie — CHANGES_REQUESTED** [review]

File: lib/Commands.cc
Comment (BewareMyPower) [reply lib/Commands.cc:445]: Good find. I think we'd better to use two overloads of `Commands::newAck` to avoid the similar misuse.

File: lib/ConsumerImpl.cc
Comment (shibd) [inline lib/ConsumerImpl.cc:1110]: Why do direct callback when readyToAck is false? 

If batchIndexAck = false, and the batch message does not have ack all messages, here also needs to wait(if ackReceiptEnabled == true)?

File: lib/AckGroupingTracker.cc
Comment (shibd) [inline lib/AckGroupingTracker.cc:97]: Here also need to callback?

File: lib/ConsumerImpl.cc
Comment (BewareMyPower) [reply lib/ConsumerImpl.cc:1110]: Yes. It has the same behavior with the Java client. See https://github.com/apache/pulsar/blob/7a99e74220bdff7a0dff3dbbf3befb2316f3a6c8/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java#L214

Even if the actual CommandAck request is not sent, the shared bit set of these `BatchMessageIdImpl` instances has already been modified, which could be treated as acknowledged.

**Review by shibd — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

**Comment by Anonymitaet** [comment]
Hi @RobertIndie, 

Just now, I discussed with @BewareMyPower on how we add docs for this PR and summarized the thoughts below. He suggests that you're an expert in this area and can provide tech inputs for this PR. Are you available? Thanks!

========================================

Here is the doc plan:

## ✅ Pulsar feature matrix

I've created ["Wait for response" in Consumer > Ack](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1784579914&range=91:91) and added versions for Pulsar and C++.

## ✅ Pulsar client docs

1. Create a topic "Wait for response" in [Client > Work with consumer](https://pulsar.apache.org/docs/next/client-libraries-consumers/#acknowledge-messages), and add a simple feature description (this info can be fetched from [here](https://github.com/apache/pulsar-client-cpp/pull/232/files#diff-258280272b1786eac0ecaafc163e189b87faa2a63ab740225da0b459f528358aR633-R639)  and code examples (Java and C++).

image

2. This doc should be added to 5 doc versions (NEXT, 2.11.x, 2.10.x, 2.9.x, 2.8.x) since this feature is available from Pulsar 2.8.x. 

- Original doc file of NEXT is [here](https://github.com/apache/pulsar-site/edit/main/docs/client-libraries-tableviews.md).
- Original doc files of all versioned docs are [here](https://github.com/apache/pulsar-site/tree/main/versioned_docs).

We can update one version first and apply the same doc changes to the rest once it gets approved. 

========================================

Feel free to reach out to me if you have questions. TIA!

**Pull request merged**
